src/libraries/System.Private.CoreLib/src/System/Globalization/Normalization.Unix.cs

```diff
@@ -41,26 +42,49 @@ internal static string Normalize(string strInput, NormalizationForm normalizatio
 
             ValidateArguments(strInput, normalizationForm);
 
-            char[] buf = new char[strInput.Length];
-
-            for (int attempts = 2; attempts > 0; attempts--)
+            char[]? toReturn = null;
+            try
             {
-                int realLen = Interop.Globalization.NormalizeString(normalizationForm, strInput, strInput.Length, buf, buf.Length);
+                int realLen = strInput.Length;
 
-                if (realLen == -1)
+                for (int attempt = 0; attempt < 2; attempt++)
                 {
-                    throw new ArgumentException(SR.Argument_InvalidCharSequenceNoIndex, nameof(strInput));
+                    if (toReturn != null)
+                    {
+                        // Clear toReturn first to ensure we don't return the same buffer twice
+                        char[] temp = toReturn;
+                        toReturn = null;
+                        ArrayPool<char>.Shared.Return(temp);
+                    }
+
+                    Span<char> buffer = realLen <= 512
+                        ? stackalloc char[512]
```

Comment:
This is running stackalloc in loop. I know that the loop is bounded in this case; but it should be still avoided.

Reply:
This stackalloc can be reached at most once since realLen will be > 512 afterwards, I'll remove it from the loop

Reply:
The Windows equivalent does this correctly. I would follow the same structure here.

Reply:
@jkotas is the concern that it's more efficient to pull the alloc out of the loop? or there's some other issue with doing it in a loop.

Reply:
Yes, allows the JIT to do better optimizations. And it will save you time on convincing [analyzer](https://github.com/dotnet/roslyn-analyzers/pull/3432) that your stackalloc in the loop is actually fine.

Reply:
Oh, I see, it overruns the stack. Never mind.